java/code/src/com/redhat/rhn/manager/contentmgmt/ContentManager.java

```diff
@@ -327,6 +344,153 @@ public static int detachSource(String projectLabel, Type sourceType, String sour
         return ContentProjectFactory.lookupProjectSource(project, sourceType, sourceLabel, user);
     }
 
+
+    /**
+     * Build a {@link ContentProject}
+     *
+     * @param projectLabel the Project label
+     * @param message the optional message
+     * @param async run the time-expensive operations asynchronously? (in the test code it is useful to run them
+     * synchronously)
+     * @param user the user
+     */
+    public static void buildProject(String projectLabel, Optional<String> message, boolean async, User user) {
+        ensureOrgAdmin(user);
+        ContentProject project = lookupProject(projectLabel, user)
+                .orElseThrow(() -> new EntityNotExistsException(projectLabel));
+        ContentEnvironment firstEnv = project.getFirstEnvironmentOpt()
+                .orElseThrow(() -> new ContentManagementException("Cannot publish  project: " + projectLabel +
+                        " with no environments."));
+        buildSoftwareSources(project, firstEnv, async, user);
+        addHistoryEntry(message, user, project);
+        firstEnv.increaseVersion();
+    }
+
+    /**
+     * Build {@link SoftwareProjectSource}s assigned to {@link ContentProject}
+     *
+     * @param project the Project
+     * @param firstEnv first Environment of the Project
+     * @param async run the time-expensive operations asynchronously?
+     * @param user the user
+     */
+    private static void buildSoftwareSources(ContentProject project, ContentEnvironment firstEnv, boolean async,
+            User user) {
+        SoftwareProjectSource leader = project.lookupSwSourceLeader()
+                .orElseThrow(() -> new ContentManagementException("Cannot publish  project: " + project.getLabel() +
+                        " with no base channel associated with it."));
```

Comment:
Whenever you need to build a `String` with spaces and parameters I always try to use `String.format` instead, it's easier to format the `String` and to read :wink: 
`String.format("Cannot publish  project: {0} with no environments.", projectLabel)`

Reply:
I think this is a matter of personal feeling. I prefer using concatenation and I've used it in all previous content management PRs.